src/m3ninx/index/segment/fst/encoding/docs/data.go

```diff
@@ -128,3 +128,83 @@ func (r *DataReader) Read(offset uint64) (doc.Document, error) {
 
 	return d, nil
 }
+
+// EncodedDataReader is a reader for the data file for encoded documents.
+type EncodedDataReader struct {
+	data []byte
+}
+
+// NewEncodedDataReader returns a new EncodedDataReader.
+func NewEncodedDataReader(data []byte) *EncodedDataReader {
+	return &EncodedDataReader{
+		data: data,
+	}
+}
+
+// Read reads a doc.EncodedDocument from a data stream starting at the specified offset.
+func (e *EncodedDataReader) Read(offset uint64) (doc.EncodedDocument, error) {
+	if offset >= uint64(len(e.data)) {
+		return doc.EncodedDocument{}, fmt.Errorf(
+			"invalid offset: %v is past the end of the data file", offset,
+		)
+	}
+
+	return doc.EncodedDocument{
+		Bytes: e.data[int(offset):],
+	}, nil
+}
+
+// EncodedDocumentReader is a reader for reading documents from encoded documents.
+type EncodedDocumentReader struct {
+	currFields []doc.Field
+}
+
+// NewEncodedDocumentReader returns a new EncodedDocumentReader.
+func NewEncodedDocumentReader() *EncodedDocumentReader {
+	return &EncodedDocumentReader{}
+}
+
+// Read reads a doc.Document from a doc.EncodedDocument. Returned doc.Document should be
+// processed before calling Read again as the underlying array pointed to by the Fields
+// slice will be updated. This approach avoids allocating a new slice with a new backing
+// array for every document processed, unlike (*DataReader).Read
+func (r *EncodedDocumentReader) Read(encoded doc.EncodedDocument) (doc.Document, error) {
+	r.currFields = r.currFields[:0]
```

Comment:
Might want to also clear out the individual fields, since probably don't want to keep ptrs to any existing slices.

e.g.
```go
for i := range r.currFields {
  r.currFields[i] = doc.Field{}
}
```